### DISCUSSION TITLE
Sorting relations

### DISCUSSION THREAD

**Discussion by fsuter** [sourcebroker/t3api#59] [Q&A] [answered]
Hi,

I have several tables related together with MM relations. I have noticed that - in the API results - the relations are not ordered as they are in the backend. Since the documentation is a bit lacking, I couldn't find if there was a way to order the relations. I have checked the examples for the news extension (https://github.com/sourcebroker/t3apinews/blob/master/Classes/Domain/Model/News.php), but I see no particular configuration for the `falMedia` field for example.

Is that actually possible at all?

**Comment by klamparski** [comment, marked as answer]
Hi @fsuter,

Just to be clear, `t3api` extension does not change default Extbase/TCA behaviour, so I guess that order is different than expected not only in API response but also in standard FE implementation (e.g. when you fetch records inside plugin from Repository). 
Thus, I think that field `foreign_sortby` might be what you are looking for: https://docs.typo3.org/m/typo3/reference-tca/main/en-us/ColumnsConfig/Type/Inline/Properties/ForeignSortby.html

Please confirm if it resolved your issue.

**Comment by fsuter** [comment]
Hi Karol. Thanks for your answer. I could find that there are several places in my project (it's a complex one) where sorting information is either not well-defined or gets lost.
Another thing regarding order, although not related to relations: I have one table with a `sorting` property. As far as I can tell, this is ignored by t3api. I looked at method `\SourceBroker\T3api\Domain\Repository\CommonRepository::findFiltered()` which seems to be central and it does not seem to use sorting anywhere. Is that right?

**Comment by klamparski** [comment]
Hi @fsuter. I guess you mean [TCA'a `ctrl.default_sortby` field](https://docs.typo3.org/m/typo3/reference-tca/10.4/en-us/Ctrl/Index.html#default-sortby). If so then you need to know that are to completely different cases. Extbase also ignores `ctrl.default_sortby` property and it uses [the one defined inside repository](https://docs.typo3.org/m/typo3/book-extbasefluid/main/en-us/6-Persistence/2a-creating-the-repositories.html#default-orderings). So sorting of the relations is done by ORM when mapping database row to model and fetching all the relations.
t3api does not change Extbase on such low level thus default behaviour is used and, as far as I remember, Extbase was using here [TCA configuration](https://docs.typo3.org/m/typo3/reference-tca/main/en-us/ColumnsConfig/Type/Inline/Properties/ForeignSortby.html) mentioned in my previous reply.
From the other hand sorting of the queried entity is done inside in query builder. As you mentioned, t3api uses `\SourceBroker\T3api\Domain\Repository\CommonRepository` for all queries and that's why applying default sorting inside repository properties is not possible. To sort the queried entity [`OrderFilter`](https://docs.typo3.org/p/sourcebroker/t3api/main/en-us/Filtering/BuiltinFilters/OrderFilter/Index.html) has to be used.

I agree that t3api should allow to apply default sorting without query parameter. Already had cases when default sorting has to be applied but for now always resolved it with query parameter inside URL (OrderFilter). But even when we introduce default sorting it will be rather done inside definition of `ApiResource` instead of using `ctrl.default_sortby` because `ctrl.default_sortby` is a plain query part (e.g. `title ASC, bodytext ASC`) so can't be passed to Extbase's query builder.

**Reply by fsuter** [reply]
Thanks for the detailed answer. I understand that applying "default_sortby" would be difficult because of its syntax, I was thinking about tables which have a sorting field defined using the "sortby" property (https://docs.typo3.org/m/typo3/reference-tca/main/en-us/Ctrl/Properties/Sortby.html#sortby). But you are right that even does not handle this and it has to be done by defining a default ordering in the repository.

I had missed the OrderFilter annotation, but indeed it's not quite what I want, since I don't want API users to have to pass a query parameter to get what should be the default sorting. In the meantime I have worked around this by xclassing `CommonRepository`(not very nice, but I needed to make progress quickly). What would be interesting would be to have something like a DefaultOrderFilter, which would apply everytime - without needing a query parameter. I could be overridden by an OrderFilter.

**Reply by klamparski** [reply]
I absolutely agree default sorting is needed. I had already projects when it would be useful but always worked it around using `OrderFilter` because of missing time for improvements of the core. FYI: Our discussion reminded me about this missing feature and I created an issue about that (#61).
